File: server_side/oidc-pam.py
#!/usr/bin/env python

# based on https://developers.onelogin.com/authentication/tools/linux-ssh-pam-module

'''
PAM module for authenticating users via a OIDC token
'''
import json
import os
import sys
import logging
import requests




def logit(data):
    '''
    Logs data to stderr and syslog
    Args:
        data (*): Data to log
    Returns: None
    '''
    logging.basicConfig(filename='/tmp/pam.log', encoding='utf-8', level=logging.DEBUG)
    data_str = str(data)
    sys.stderr.write('%s\n' % data_str)
    logging.debug(data_str)

def pam_sm_setcred(pamh, _flags, _argv):
    '''
    Default
    '''
    return pamh.PAM_SUCCESS
 
def pam_sm_acct_mgmt(pamh, _flags, _argv):
    '''
    Default
    '''
    return pamh.PAM_SUCCESS
 
def pam_sm_open_session(pamh, _flags, _argv):
    '''
    Default
    '''
    return pamh.PAM_SUCCESS
 
def pam_sm_close_session(pamh, _flags, _argv):
    '''
    Default
    '''
    return pamh.PAM_SUCCESS
 
def pam_sm_chauthtok(pamh, _flags, _argv):
    '''
    Default
    '''
    return pamh.PAM_SUCCESS


def pam_sm_authenticate(pamh, _flags, _argv):
    '''
    Authenticates a user via an OIDC token
    '''
    # Load config file and build access token
    try:
        config_dpath = os.path.dirname(os.path.realpath(__file__))
        config_fpath = os.path.join(config_dpath, 'oidc-pam.json')
        config_fd = open(config_fpath, 'r')
        config = config_fd.read()
        config_fd.close()
        config = json.loads(config)
    except Exception as error:
        logit('Error loading configuration: %s' % error)
        return pamh.PAM_AUTH_ERR

    # get user&token
    try:
        user = pamh.get_user(None)
        if user is None:
            return pamh.PAM_USER_UNKNOWN
        access_token = pamh.authtok
#        if len(access_token)<20: #cannot be token, should be wrong password
#            return pamh.PAM_AUTH_ERR
        if len(access_token)>5:
            next_token_part = pamh.conversation(pamh.Message(pamh.PAM_PROMPT_ECHO_OFF,'Next: ')).resp
            while (next_token_part != 'token_end') and (next_token_part!=''):
                access_token = access_token + next_token_part                 
                next_token_part = pamh.conversation(pamh.Message(pamh.PAM_PROMPT_ECHO_OFF,'Next: ')).resp
        if access_token is None:
            logit('empty access_token token')
            return pamh.PAM_AUTH_ERR
    except pamh.exception as error:
        return error.pam_result

    # todo: check user same as in token
    try:
        url = config['introspection_url']
        logit(access_token)
        data = {'token': access_token.strip(),'client_id': config['client_id'], 'client_secret':config['client_secret']}
        response = requests.post(url, data = data)
        if response.status_code != requests.status_codes.codes.ok:
            logit('Error checking introspecting token, server returned %d %s' % response.status_code, response.text)
            return pamh.PAM_AUTH_ERR
        token_info=response.json()
        if token_info['active']!= True:
            logit('Error checking introspecting token, token %s invalid, server response: %s' %(access_token, response.text))
            return pamh.PAM_AUTH_ERR            
        logit(response.json())
    except Exception as error:
        logit('Error introspecting token %s, error: %s' % (access_token,error))
        return pamh.PAM_AUTH_ERR

    return pamh.PAM_SUCCESS

File: server_side/c/oidc-pam.c
#include <stdio.h>
#include <string.h>
#include <security/pam_appl.h>
#include <security/pam_modules.h>
#include <stdlib.h>

#include "config.h"
#include "auth.h"
#include "log.h"

/* expected hooks */
PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc, const char **argv) {
    return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_close_session(pam_handle_t *pamh, int flags, int argc, const char **argv) {
    return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_chauthtok(pam_handle_t *pamh, int flags, int argc, const char **argv) {
    return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char **argv) {
    return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc, const char **argv) {
    return PAM_SUCCESS;
}

int conversation(pam_handle_t *pamh, const char *prompt, char **response) {
    struct pam_message msg;
    struct pam_conv *conv;
    struct pam_response *pResp;
    const struct pam_message *pMsg = &msg;
    msg.msg_style = PAM_PROMPT_ECHO_OFF;
    msg.msg = prompt;
    int rc = pam_get_item(pamh, PAM_CONV, (const void **) &conv);
    if (rc != PAM_SUCCESS) {
        logstr("error in pam_get_item\n");
        return rc;
    }
    rc = conv->conv(1, (const struct pam_message **) &pMsg, &pResp, conv->appdata_ptr);
    if (rc != PAM_SUCCESS || pResp == NULL) {
        logstr("error in conv\n");
        return PAM_CONV_ERR;
    }
    if (pResp[0].resp == NULL) {
        logstr("empty response\n");
        free(pResp);
        return PAM_AUTH_ERR;
    }

    strcpy(*response, pResp[0].resp);
    free(pResp[0].resp);
    pResp[0].resp = NULL;
    free(pResp);
    return 0;
}

int get_access_token(pam_handle_t *pamh, int use_first_pass, char **access_token) {
    if (use_first_pass) {
        const char *saved_token;
        pam_get_authtok(pamh, PAM_AUTHTOK, &saved_token, NULL);
        strcpy(*access_token, saved_token);
    } else {
        int res = conversation(pamh, "Passcode or token: ", access_token);
        if (res != 0) {
            logstr("error in getting passcode or token\n");
            return res;
        }
    }
    if (strlen(*access_token) < 20) { // assuming this is a password and should be processed in other module
        pam_set_item(pamh, PAM_AUTHTOK, *access_token);
        return PAM_AUTH_ERR;
    }

    char *next_token_part = (char *) malloc(10000 * sizeof(char));
    int res = conversation(pamh, "Next: ", &next_token_part);
    if (res != 0) {
        logstr("error in getting next token part\n");
        free(next_token_part);
        return res;
    }
    while (strlen(next_token_part) > 0 && strcmp(next_token_part, "token_end") != 0) {
        strcat(*access_token, next_token_part);
        res = conversation(pamh, "Next: ", &next_token_part);
        if (res != 0) {
            logstr("error in getting next token part\n");
            free(next_token_part);
            return res;
        }
    }
    if (strlen(*access_token) == 0) {
        logstr("empty access token\n");
        free(next_token_part);
        return PAM_AUTH_ERR;
    }

    free(next_token_part);
    return 0;
}

PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc, const char **argv) {
    int res = parse_config(argv[0], &config);
    if (res != 0) {
        logstr("cannot parse config file\n");
        return PAM_AUTH_ERR;
    }

    int use_first_pass = argc == 2 && strcmp(argv[1], "use_first_pass") == 0;
    // get user
    const char *pUsername;
    int retval = pam_get_user(pamh, &pUsername, NULL);
    if (retval != PAM_SUCCESS) {
        logstr("unknown user\n");
        return PAM_USER_UNKNOWN;
    }
    // get token
    char *access_token = (char *) malloc(10000 * sizeof(char));
    retval = get_access_token(pamh, use_first_pass, &access_token);
    if (retval != 0) {
        cJSON_Delete(config.parsed_object);
        free(access_token);
        return retval;
    }

    oidc_token_content_t token_content;
    res = verify_token(access_token, &token_content);
    if (res != 0) {
        logit("error introspecting token: %s\n", access_token);
        cJSON_Delete(config.parsed_object);
        free(access_token);
        return PAM_AUTH_ERR;
    }

    int token_ok = 1;
    if (strncmp(token_content.user, pUsername, 3) != 0) {
        logit("error checking username, token: %s, user: %s\n", access_token, pUsername);
        token_ok = 0;
    }

    ///TODO: Figure out 2fa

    cJSON_Delete(config.parsed_object);

    free(access_token);
    free(token_content.user);

    if (token_ok == 0) {
        return PAM_AUTH_ERR;
    }

    return PAM_SUCCESS;
}